### PR TITLE
Changed spark sharding of queryname sorted bams to better handle secondary and supplementary reads

### LINKED ISSUE #4441: Spark custom partitioning of mates doesn't handle secondary or supplementary reads
When we ingest a queryname sorted bam in spark using `ReadsSparkSource.putPairsInSamePartition()`, we shift reads between partitions in an attempt to keep reads together with their mates. Unfortunately the code currently makes the assumption that there are only two mates in a pair and thus only shuffles around at most one read per partition. This assumption is not entirely correct, as if there are secondary or supplementary reads by the same name they could end up split between partitions. 

Looking at the tie breaking in SortSam, it seems like this approach could easily result in reads being separated from their primary mates under certain circumstances in the bam.

### PR DESCRIPTION
**Pull request by jamesemery**
Because the sorting order makes it difficult to know if the reads you are holding are in the grouped properly at the front of a shard, we must serialize all the starting reads grouped by their name and append them at the backs of shards appropriately. 

Fixes #4441

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 1]
```diff
@@ -217,26 +218,53 @@ public boolean accept(Path path) {
...
+                // If a read spans multiple partitions we expect its name to show up multiple times and we don't expect this to work properly
+                if (groupName != null && group.get(0).getName().equals(groupName)) {
+                    throw new GATKException(String.format("The read name '%s' appeared across multiple partitions this could indicate there was a problem " +
+                            "with the sorting or that it has too many groups, check that the file is queryname sorted and consider decreasing the number of paritions", groupName));
+                }
+                groupName =  group.get(0).getName();
+            }
+        }
+
         // Shift left, so that each partition will be joined with the first read from the _next_ partition
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: comment out of date now

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 2]
```diff
@@ -217,26 +218,53 @@ public boolean accept(Path path) {
...
-                (FlatMapFunction2<Iterator<GATKRead>, Iterator<GATKRead>, GATKRead>) (it1, it2) -> {
+                (FlatMapFunction2<Iterator<GATKRead>, Iterator<List<GATKRead>>, GATKRead>) (it1, it2) -> {
             PeekingIterator<GATKRead> current = Iterators.peekingIterator(it1);
-            // skip the first read in the _current_ partition if it is the second in a pair since it will be handled in the previous partition
-            if (current.hasNext() && current.peek() != null && current.peek().isSecondOfPair()) {
-                current.next();
+            String firstName = current.peek().getName();
+            // Make sure we don't remove reads from the first partition
+            if (!firstGroupInBam.equals(firstName)) {
+                // skip the first readGroup in the _current_ partition if it is the second in a pair since it will be handled in the previous partition
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: `readGroup` -> `group of reads with the same name` or `read name group`

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 3]
```diff
@@ -217,26 +218,53 @@ public boolean accept(Path path) {
...
+                    firstGroup.add(current.next());
+                }
+                return Iterators.singletonIterator(firstGroup);
+                })
                 .collect();
+
+        // Checking for pathological cases (read name groups that span multiple partitions)
+        String groupName = null;
+        for (List<GATKRead> group : firstReadNamesInEachPartition) {
+            if (group!=null && !group.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: at the edges, we don't have anything for the n+1th partition after we shifted it left so we have to just ensure we are doing something rational there. Also it's not clear to me there aren't conditions under which there might be empty partitions produced by spark.

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 4]
```diff
@@ -217,26 +218,53 @@ public boolean accept(Path path) {
...
+                // If a read spans multiple partitions we expect its name to show up multiple times and we don't expect this to work properly
+                if (groupName != null && group.get(0).getName().equals(groupName)) {
+                    throw new GATKException(String.format("The read name '%s' appeared across multiple partitions this could indicate there was a problem " +
+                            "with the sorting or that it has too many groups, check that the file is queryname sorted and consider decreasing the number of paritions", groupName));
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: done

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java [hunk 5]
```diff
@@ -217,26 +218,53 @@ public boolean accept(Path path) {
...
         }
         int numPartitions = reads.getNumPartitions();
-        // Find the first read in each partition
-        List<GATKRead> firstReadInEachPartition = reads
-                .mapPartitions((FlatMapFunction<Iterator<GATKRead>, GATKRead>) it -> Iterators.singletonIterator(it.next()))
+        final String firstGroupInBam = reads.first().getName();
+        // Find the first group in each partition
+        List<List<GATKRead>> firstReadNamesInEachPartition = reads
+                .mapPartitions((FlatMapFunction<Iterator<GATKRead>, List<GATKRead>>) it ->
+                { PeekingIterator<GATKRead> current = Iterators.peekingIterator(it);
```
[the remark this answers is not in the document]
Comment (jamesemery) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: done

File: src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lbergelson) [reply src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java]: ah, yes, good point

**Pull request merged**